airbyte-tests/src/main/java/io/airbyte/test/airbyte_test_container/AirbyteTestContainer.java

```diff
@@ -0,0 +1,276 @@
+    // todo (cgardens) - assumes port 8001 which is misleading since we can start airbyte on other
+    // ports. need to make this configurable.
+    final AirbyteApiClient apiClient = new AirbyteApiClient(
+        new ApiClient().setScheme("http")
+            .setHost("localhost")
+            .setPort(8001)
+            .setBasePath("/api"));
+
+    final HealthApi healthApi = apiClient.getHealthApi();
+
+    ApiException lastException;
+    int i = 0;
+    while (true) {
+      try {
+        healthApi.getHealthCheck();
+        break;
+      } catch (ApiException e) {
+        lastException = e;
+        LOGGER.info("airbyte not ready yet. attempt: {}", i);
+      }
+      if (i == 10) {
+        throw new IllegalStateException("Airbyte took too long to start. Including last exception.", lastException);
+      }
+      Thread.sleep(5000);
+      i++;
+    }
+  }
+
+  private void serviceLogConsumer(DockerComposeContainer<?> composeContainer, String service) {
+    composeContainer.withLogConsumer(service, logConsumer(customServiceLogListeners.get(service)));
+  }
+
+  /**
+   * Exposes logs generated by docker containers in docker compose temporal test container.
+   *
+   * @param customConsumer - each line output by the service in docker compose will be passed ot the
+   *        consumer. if null do nothing.
+   * @return log consumer
+   */
+  private Consumer<OutputFrame> logConsumer(Consumer<String> customConsumer) {
+    return c -> {
+      if (c != null && c.getBytes() != null) {
+        final String log = new String(c.getBytes());
+        if (customConsumer != null) {
+          customConsumer.accept(log);
+        }
+        LOGGER.info(log.replace("\n", ""));
+      }
+    };
+  }
+
+  /**
+   * This stop method will delete any underlying volumes for the docker compose setup.
+   */
+  public void stop() {
+    if (dockerComposeContainer != null) {
+      dockerComposeContainer.stop();
+    }
+  }
+
+  /**
+   * This method is hacked from {@link org.testcontainers.containers.DockerComposeContainer#stop()} We
+   * needed to do this to avoid removing the volumes when the container is stopped so that the data
+   * persists and can be tested against in the second run
+   */
+  public void stopRetainVolumes() {
+    if (dockerComposeContainer == null) {
+      return;
+    }
+
+    try {
+      stopRetainVolumesInternal();
+    } catch (InvocationTargetException | IllegalAccessException | NoSuchMethodException | NoSuchFieldException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @SuppressWarnings("rawtypes")
+  private void stopRetainVolumesInternal() throws InvocationTargetException, IllegalAccessException, NoSuchMethodException, NoSuchFieldException {
+    final Class<? extends DockerComposeContainer> dockerComposeContainerClass = dockerComposeContainer.getClass();
+    try {
+      final Field ambassadorContainerField = dockerComposeContainerClass.getDeclaredField("ambassadorContainer");
```

Comment:
what is the `ambassadorContainerField` and what is the `SocatContainer`?

Reply:
this is the reflection stuff that subodh figured out so that we could spin down a docker compose test container without deleting volumes. i'm just moving the code. i'd have to dig a bit deeper to remember how this works.